### PR TITLE
Type conversion in C-lisp

### LINKED ISSUE #76: Quantized types and casting in C-Lisp
Support for more types and casts between them was introduced in Brilisp in #68. Now the same capabilities need to be exposed in C-Lisp.

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
All `gen_*_expr` functions have been refactored to return the result type along with the instructions list and the result symbol. This is required since we now need to track the types of all intermediate symbols, and having a symbol table just for that is not the most elegant solution.

Fix #76

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
As discussed, perhaps we should create a simple class that wraps the instructions list, name and type for a symbol. It may help new contributors to the project if the constructor of such a class reports an error when the return value is not of the right form.

File: src/backend/c-lisp.py [hunk 1]
```diff
@@ -178,7 +148,7 @@ def gen_while_stmt(self, stmt):
                 "break",
             )
         ]
-        cond_expr_instr, cond_sym = self.gen_expr(stmt[1])
+        cond_expr_instr, cond_sym, _ = self.gen_expr(stmt[1])
```
Comment (chsasank) [inline src/backend/c-lisp.py:151]: put expression in a class

File: src/backend/c-lisp.py [hunk 2]
```diff
@@ -314,182 +323,289 @@ def get_literal_type(self, expr):
...
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] in cls.cast_ops
+
+    def compile(self, expr):
+        opcode = expr[0]
+        if len(expr) != 3:
+            raise CodegenError(f"`{opcode}` takes exactly 2 operands")
+
+        # TODO: Type checking
```
Comment (chsasank) [inline src/backend/c-lisp.py:599]: Add tests and type checking. I want to merge this.

Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py:599]: On it

Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py:599]: Almost done. Only inttoptr and ptrtoint remain

Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Tests for all cast operators and type-checking for casting is done.

**Comment by chsasank** [comment]
PR looks good, can I merge?

**Pull request merged**